cfn-resources/stream-instance/cmd/resource/resource.go

```diff
@@ -0,0 +1,318 @@
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/progressevent"
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/validator"
+	"go.mongodb.org/atlas-sdk/v20231115007/admin"
+)
+
+func setup() {
+	util.SetupLogger("mongodb-atlas-stream-instance")
+}
+
+var CreateRequiredFields = []string{constants.Name, constants.GroupID, constants.StreamConfig, constants.DataProcessRegion}
+var ReadRequiredFields = []string{constants.Name, constants.GroupID}
+var UpdateRequiredFields = []string{constants.Name, constants.GroupID, constants.DataProcessRegion}
+var DeleteRequiredFields = []string{constants.Name, constants.GroupID}
+var ListRequiredFields = []string{constants.GroupID}
+
+const Kafka = "Kafka"
+const Cluster = "Cluster"
+
+func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup()
+	util.SetDefaultProfileIfNotDefined(&currentModel.Profile)
+	if errEvent := validator.ValidateModel(CreateRequiredFields, currentModel); errEvent != nil {
+		return *errEvent, nil
+	}
+
+	client, handlerError := util.NewAtlasV2OnlyClientLatest(&req, currentModel.Profile, true)
+	if handlerError != nil {
+		_, _ = logger.Warnf("CreateMongoDBClient error: %v", handlerError)
+		return *handlerError, errors.New(handlerError.Message)
+	}
+
+	streamInstanceCreateReq := newStreamsTenant(currentModel)
+
+	atlasV2 := client.AtlasSDK
+
+	createdStreamInstance, resp, err := atlasV2.StreamsApi.CreateStreamInstance(context.Background(), *currentModel.GroupId, streamInstanceCreateReq).Execute()
+	if err != nil {
+		return handleError(resp, constants.CREATE, err)
+	}
+
+	currentModel.Id = createdStreamInstance.Id
+
+	return handler.ProgressEvent{
+		OperationStatus: handler.Success,
+		Message:         "Create Complete",
+		ResourceModel:   currentModel,
+	}, nil
+}
+
+func Read(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup()
+	util.SetDefaultProfileIfNotDefined(&currentModel.Profile)
+	if errEvent := validator.ValidateModel(ReadRequiredFields, currentModel); errEvent != nil {
+		return *errEvent, nil
+	}
+
+	client, handlerError := util.NewAtlasV2OnlyClientLatest(&req, currentModel.Profile, true)
+	if handlerError != nil {
+		_, _ = logger.Warnf("CreateMongoDBClient error: %v", handlerError)
+		return *handlerError, errors.New(handlerError.Message)
+	}
+
+	atlasV2 := client.AtlasSDK
+
+	streamInstance, resp, err := atlasV2.StreamsApi.GetStreamInstance(context.Background(), *currentModel.GroupId, *currentModel.Name).Execute()
+	if err != nil {
+		return handleError(resp, constants.READ, err)
+	}
+
+	currentModel.Id = streamInstance.Id
+	currentModel.Hostnames = streamInstance.GetHostnames()
+	currentModel.DataProcessRegion = newModelDataRegion(streamInstance.DataProcessRegion)
+	currentModel.StreamConfig = newModelStreamConfig(streamInstance.StreamConfig)
+	currentModel.Connections = newModelConnections(streamInstance.Connections)
```

Comment:
would it be worth extract this to a method? Such as `newCFNStreamInstance`. I believe it can be used in the for the List method as well, and can be helpful for adding unit test for the logic of passing from SDK model to CFN model.

Reply:
List and Read are a bit different when tranforming from SDK to CFN models(list sets more attributes). I would prefer to keep it as is so it's very clear which attributes are being set. Although thanks to this comment, I reused some methods, such as `newModelStreamConfig` in the List method so code is cleaner

Reply:
since they are the same model being populated in both operations, is there a reason we want to populate more fields for the list operation? or less in the read.

Reply:
The reason it's different is because:
- Model for read has the InstanceName attribute already set (needed to be able to read specific StreamInstance)
- List has to create a new model for each one of the results, so attributes already set in the `currentModel` have to be set in the new model objects(such as Profile and ProjectId)